### ISSUE TITLE
[FEA]: Capture potential segmented reduce transform API and performance

### ISSUE THREAD

**Issue by gevtushenko** [NVIDIA/cccl#9447]
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

CUB

### Is your feature request related to a problem? Please describe.

One of the use cases for hierarchical parallelism is RMS Norm https://github.com/NVIDIA/cccl/issues/8358#issuecomment-4237875945. That said, we can provide much simpler interface for RMS Norm, something along the lines of `cub::DeviceSegmentedReduceTransform` that'd use hierarchical parallelism internally, but not necesserily expose it through the interface. 

### Describe the solution you'd like

This issue can be closed by filing an issue or discussion that captures:
- segmented reduce API
- list of algorithms that can be implemented with it (RMS Norm, etc)
- performance of preliminary implementation compared to other solutions (PyTorch, cuTile, etc)

If sufficient number of use cases around this interface accumulate over time, we'll consider implementing it.

**Comment by NaderAlAwar** [comment]
## List of algorithms that can be implemented

### Normalization algorithms

Using PyTorch as an example, this includes [RMSNorm](https://docs.pytorch.org/docs/2.12/generated/torch.nn.RMSNorm.html), [LayerNorm](https://docs.pytorch.org/docs/2.12/generated/torch.nn.LayerNorm.html), [Softmax](https://docs.pytorch.org/docs/2.12/generated/torch.nn.Softmax.html), [Normalize](https://docs.pytorch.org/docs/2.12/generated/torch.nn.functional.normalize.html), [renorm](https://docs.pytorch.org/docs/2.12/generated/torch.renorm.html), and [weight_norm](https://docs.pytorch.org/docs/2.12/generated/torch.nn.utils.weight_norm.html). See [RMSNorm/LayerNorm kernel](https://github.com/pytorch/pytorch/blob/56964c25c21235cf3a06679d2e400195087f64fb/aten/src/ATen/native/cuda/layer_norm_kernel.cu#L1122-L1208), other algorithms use different kernels, but they should still be expressible as `ReduceTransform`.

### Quantization algorithms

Quantization. FP8 activation quantization is one example that should fit the `ReduceTransform` pattern. The reduction is a max over each quantization block, which is used to compute both a float32 inverse scale and a one-byte E8M0 scale. The transform then maps each input element to one quantized FP8 output byte.

The main API complication is that the reduce prologue produces metadata that must be stored: one E8M0 scale byte per segment. The float32 inverse scale is needed by the following transform, but does not necessarily need to be written to memory. So the API needs a way to expose/store selected prologue outputs. One possible approach is a transform output iterator that writes the E8M0 byte and discards the temporary inverse-scale value.

There may also be performance concerns because the CUB version relies heavily on custom iterators and functors, which can make vectorization/TMA harder.

Implementation-side note: it may be useful to store the prologue output as soon as it is computed, so the metadata store can overlap with the transform work, instead of delaying all stores until the end and creating tail effects.

See kernel: https://github.com/NVIDIA/TensorRT-LLM/blob/07180493c8b6ef12688cfb152c21a5f7200e26db/cpp/tensorrt_llm/kernels/quantization.cuh#L758-L902

## `SegmentedReduceTransform` API

The API should express the algorithms mentioned above. The most common use cases require fixed-size segments, mainly because LLMs process tokens in batches and each token is represented by a fixed-size vector. So the initial implementation does not need to support variable-size segments. The API also needs support for storing the output of the reduce/prologue portion,  because some algorithms produce useful per-segment metadata in addition to the final per-element output.

The proposed API looks like a fixed-size `SegmentedReduce` API with some additions:

```cpp
  cub::DeviceSegmentedReduceTransform::ReduceTransform(
    d_in, d_out, num_segments, segment_size,
    /* reduction_op */ [](float l, float r) { return l + r; },
    initial_value,
    /* transform_op */ [](float sum, float val) { return val / sum; });
```

A second overload allows specifying an additional output iterator for the per-segment reduction/prologue output. Both overloads can use the same kernel internally; the first form can use a `discard_iterator` for the reduce output.


```cpp
  cub::DeviceSegmentedReduceTransform::ReduceTransform(
    InputIteratorT d_in,
    ReduceOutputIteratorT d_reduce_out,
    TransformOutputIteratorT d_transform_out,
    cuda::std::int64_t num_segments,
    int segment_size,
    ReductionOpT reduction_op,
    T initial_value,
    TransformOpT transform_op,
    EnvT env = {});
```

An important detail is that the reduction operator should look like a normal Reduce/SegmentedReduce operator. It should not take into expose thread hierarchies to the user.

There are other API considerations not specific to `ReduceTransform` that need to be taken into account:

1. For RMSNorm, the reduce op requires the input to be squared using a transform iterator but the transform needs the original input. For performance reasons, we should only load from the global memory buffer into shared memory once per segment. We could potentially apply the transform so the reduce op can use it when loading from shared memory, but that requires redesigning how iterators work
2. For RMSNorm, we need to load the weights which are used once per segment. We need some kind of iterator that repeats the weights buffer multiple times so that it has a logical size that matches the main input. Essentially, we need to "repeat" the weights `num_segments` times.

### Performance measurements

<img width="2769" height="1257" alt="Image" src="https://github.com/user-attachments/assets/b1b5e14b-20fd-4432-8ef1-0bef6025e7bd" />

<img width="2769" height="1257" alt="Image" src="https://github.com/user-attachments/assets/753d9bd9-7f35-4e49-94e1-d03a422cedd6" />

This compares our implementation, C++ hierarchical POC final, to others. For both GPUs, we beat other implementations most of the time (`DeviceTransform` does not actually implement RMSNorm, it only loads and stores the same amount of data, and functions as an upper bound of what can be achieved). We are still worse for small hidden/segment sizes and the largest size on the B200. There are three main optimization opportunities left:

1. Use multiple segments per block for small segment sizes.
2. Use clusters for larger segment sizes instead of having one block reduce an entire segment.
3. Use pipelining or warp specialization. This would involve having one warp stage part of the segment, while other warps reduce it

The hierarchical parallelism implementation lives in my branch here https://github.com/NaderAlAwar/cccl/tree/hierarchical-parallelism commit 8b5f210124618f53993c5bf5115fad34dfe2c719. The benchmarks can be found under https://github.com/NaderAlAwar/cccl/tree/hierarchical-parallelism/cub/benchmarks/bench/hierarchical/rmsnorm. and the hierarchical parallelism benchmark showing our interface can be found here https://github.com/NaderAlAwar/cccl/blob/hierarchical-parallelism/cub/benchmarks/bench/hierarchical/rmsnorm/hierarchical_rmsnorm.cu. Note that this does not currently use the proposed API above, as it exposes thread hierarchies to the user. It is expected that the `SegmentedReduceTransform` API will be higher level and built on top of the current implementation.

Note that I have not benchmarked the fused add variant, which is very widely used in LLM inference, but I expect similar performance results

**Comment by NaderAlAwar** [comment]
### Help Us Prioritize Parameters for `DeviceSegmentedReduceTransform`

We are planning the implementations for `DeviceSegmentedReduceTransform` and we'd like your input to help prioritize which variants we should focus on.

Please copy & paste the form below into a new comment and fill it out in order to describe **which `ReduceTransform` features or variants are most important to you**. Your responses will help us understand common use cases, preferred behaviors, and performance requirements, so we can prioritize the development of the most relevant versions.

<details>
<summary> 🗒️ <b>Segmented ReduceTransform Algorithm Specification Form (click to expand)</b></summary>

````markdown

# 🧮 Segmented ReduceTransform Algorithm Specification Form

Please fill out the form below to describe which **ReduceTransform algorithm parameters or variants** you need.  
Use the checkboxes (`[x]`) to select the features or behaviors you’d like supported (multiple selections are fine). You can also add clarifying notes inline.

**Use case:**  
```
If possible, briefly describe how or where you plan to use the Segmented ReduceTransform algorithm.
```

---

## ⚙️ Basic Algorithmic Properties

**Fixed segment size vs Variable segment size* *(you can select both if multiple modes are acceptable):
 - [ ] Fixed size (every segment has the same size)
 - [ ] Variable size (different segments could have different segment sizes

**Save reduce output**
**Would you like to be able to save the result of the segmented reduce part?**
- [ ] Yes
- [ ] No

**Reduce Operators**:
**What operators would you expect to use on the segmented reduce part?**
- [ ] Sum
- [ ] Max
- [ ] Custom ☐ (please add a decription)

## 💾 Data Characteristics

**Common data types** (select all that apply):
- [ ] **Built-in / fundamental types**: standard integer and floating-point types (`uint8_t`, `int32_t`, `float`, `double`, etc.)
- [ ] **Extended floating-point types**: fp16, bfloat16, ...
- [ ] **Custom data types**
- [ ] **Specific subset only**: I primarily care about a few specific types: ☐ (please list them)

**Will input and output data types differ?**:
- [ ] Yes
- [ ] No

## 📏 Typical Problem Sizes

**Common number of total input elements (i.e., the sum of segment sizes over all segments) (`N`):**
- [ ] 1 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [ ] Variable / depends on workload
- [ ] Custom: ☐ (please specify exact or range values)

**Common upper bound of segment size (i.e., the larges segment size across all segments):**
- [ ] 1 – 9
- [ ] 10 – 99
- [ ] 100 – 999
- [ ] 1 000 – 9 999
- [ ] 10 000 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [ ] Variable / depends on workload
- [ ] Custom: ☐ (please specify exact or range values)
````

</details>
